GH-1500: As a smith I want to migrate the HLC Builder to use LSP Server (Part 3)

mmews-n4 opened the pull request:
see #1500

szarnekow commented on plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/server/N4JSProjectManager.java:
```diff
@@ -57,12 +59,12 @@ public void initialize(ProjectDescription description, IProjectConfig pProjectCo
 			uris.addAll(srcFolderEx.getAllResources());
 		}
-		return doBuild(uris, Collections.emptyList(), Collections.emptyList(), cancelIndicator);
+		return doBuild(new LinkedList<>(uris), Collections.emptyList(), Collections.emptyList(), cancelIndicator);
```
https://stackoverflow.com/questions/322715/when-to-use-linkedlist-over-arraylist-in-java/42529652#42529652

szarnekow commented on plugins/org.eclipse.n4js.utils/src/org/eclipse/n4js/utils/io/FileUtils.java:
```diff
@@ -157,11 +157,7 @@ public static Path createTempDirectory(String prefix) {
-		} catch (IOException e) {
-			return null;
-		}
+		return child.toPath();
```
👍

mor-n4 commented on plugins/org.eclipse.n4js/src/org/eclipse/n4js/validation/validators/packagejson/N4JSProjectSetupJsonValidatorExtension.xtend:
```diff
@@ -370,7 +360,8 @@ public class N4JSProjectSetupJsonValidatorExtension extends AbstractJSONValidato
 	/** IDEBUG-266 issue error warning on cyclic dependencies. */
-	@Check
+	// TODO: GH-1500: Check if cyclic dependencies are already found in BuildManager#sortByDependencies()
+	// @Check // Commented out because there are performance issues.
```
Is it intentional to merge this TODO to master?

mmews-n4 replied:
This should be done when working on #1525. I'll change the issue id.

mor-n4 commented on plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/server/N4JSProjectManager.java:
```diff
@@ -151,6 +150,14 @@ public XResult doBuild(List<URI> dirtyFiles, List<URI> deletedFiles, List<Delta>
+		int size1 = dirtyFiles.size();
+		int size2 = uris.size();
+		if (size1 != size2) {
+			// TODO: check that dirtyFiles has no duplicates!
```
TODO?

mor-n4 commented on plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/server/N4JSProjectManager.java:
```diff
@@ -189,9 +196,9 @@ public XBuildRequest newBuildRequest(List<URI> changedFiles, List<URI> deletedFi
-		}
+		// if (projectConfig.getPath().segmentsList().contains(N4JSGlobals.NODE_MODULES)) {
+		// result.setAfterValidate((uri, issues) -> false);
+		// }
```
Can this be deleted?

mor-n4 commented on plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/server/ReflectionUtils.java:
```diff
@@ -0,0 +1,36 @@
+/**
+ *
+ */
+public class ReflectionUtils {
```
Should be moved to bundle `org.eclipse.n4js.utils`.

mmews-n4 replied:
Ok. Alternatively `ReflectionUtils` will be removed.

mor-n4 commented on tests/org.eclipse.n4js.hlc.integrationtests/src/org/eclipse/n4js/hlc/integrationtests/CliJarWithVerdaccioTest.java:
```diff
@@ -0,0 +1,109 @@
+/**
+ * The tests in this class test n4js-cli npm.
+ */
+public class CliJarWithVerdaccioTest extends AbstractCliJarTest {
```
would it make sense to already prepare an abstract class `AbstractCliJarWithVerdaccioTest` to separate that from the actual test cases in this class?

mor-n4 commented on tests/org.eclipse.n4js.xpect.ui.tests/testdata/project_circular_dependencies/packageA.json.xt:
```diff
@@ -31,7 +31,7 @@
 
 {
-	// XPECT errors --> "The project configuration contains a dependency cycle of projects [A] -> C -> B -> [A]." at ""A""
+	// TODO: GH-1500 XPECT FIXME errors --> "The project configuration contains a dependency cycle of projects [A] -> C -> B -> [A]." at ""A""
```
Since you won't fix this as part of `GH-1500`, please create a sub task for the cycle detection improvements and link its task ID here (or maybe use the next general LSP task for that).
(also applies to 3 other files in same folder)

mmews-n4 commented on plugins/org.eclipse.n4js/src/org/eclipse/n4js/binaries/BinariesLocatorHelper.java:
```diff
@@ -164,6 +164,63 @@ public Path findYarnPath() {
+		javaPathCandidate = new File(BinariesConstants.BUILT_IN_DEFAULT_JAVA_PATH).toPath();
+
+		return javaPathCandidate;
+	}
```
The whole `binaries` package and all related sources need to be refactored in a later step, since the lsp won't need it anymore.

mmews-n4 commented on plugins/org.eclipse.n4js/src/org/eclipse/n4js/binaries/BinariesLocatorHelper.java:
```diff
@@ -164,6 +164,63 @@ public Path findYarnPath() {
+				tryGetEnvOrSystemVariable(BinariesConstants.DEFAULT_JAVA_PATH_VM_ARG));
+		if (javaPathCandidate != null) {
+			System.out.println("User specified default java path will be used: '" + javaPathCandidate
+					+ ".' based on the '" + BinariesConstants.DEFAULT_JAVA_PATH_VM_ARG + "' VM argument.");
```
This was for testing purposes only. I reverted it to use the logger again.

mmews-n4 commented on plugins/org.eclipse.n4js/src/org/eclipse/n4js/validation/validators/packagejson/AbstractJSONValidatorExtension.java:
```diff
@@ -264,8 +268,22 @@ protected static String getJSONValueDescription(EClass valueClass) {
+			return false;
+		}
+
+		// this validator extension only applies to package.json files located in the root of a project
```
Right, this is not named something like `PackageJsonValidatorExtension`. But actually the subclasses of this class are only `N4JSProjectSetupJsonValidatorExtension` and `PackageJsonValidatorExtension`. All of these three classes are located in the n4js main bundle (meaning not in the n4js.json bundle), so the restriction to limit the scope of these validators to package.json files in the root folder only is ok, imho. Nevertheless, the abstract class could be renamed... I'll do that.

mmews-n4 replied:
I checked again and the name of this class should be `AbstractPackageJsonValidatorExtension`. Its validations refer not to generic `.json` files but `package.json`s only. The validation of generic `.json` files is still active on this branch and is implemented in the bundle `org.eclipse.n4js.json` in the class `JSONValidator`.

Resolution: I rename this file.

mmews-n4 commented:
@jpilgrim Regarding documentation:
I updated the chapter about the HLC in the ide.spec document. I also removed those few and outdated references to the HLC from the design document.

mmews-n4 commented:
Builds green. Merging.
